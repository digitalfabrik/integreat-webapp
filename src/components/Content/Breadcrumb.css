@import url('../Helper/Helper.css');

.breadcrumbs {
  margin-bottom: 20px;
}

.separator::after {
  color: #e8e8e8;
  font-size: 22px;
  content: ' > ';
  user-select: none;
}

.level {
  color: #9c9c9c;
  font-size: 20px;
  user-select: none;
}

.breadcrumb {
  @mixin remove-a;
}

.breadcrumb:last-child .level {
  color: black;
}
