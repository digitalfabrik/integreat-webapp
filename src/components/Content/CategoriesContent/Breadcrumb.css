@import url('../../Common/Helper.css');
@import url('../../Common/Colors.css');

.breadcrumbs {
  white-space: nowrap;
  overflow-x: auto;
}

.separator::after {
  color: var(--text-decoration-color);
  font-size: 16px;
  content: ' > ';
}

.level {
  color: var(--text-secondary-color);
  font-size: 15px;
}

.breadcrumb {
  @mixin remove-a;
}
