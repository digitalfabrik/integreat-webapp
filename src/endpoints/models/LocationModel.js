import { isEmpty } from 'lodash/lang'
import { find } from 'lodash/collection'

const IGNORED_PREFIXES = ['stadt', 'kreis', 'landkreis']

export default class LocationModel {
  constructor (name, code, live) {
    this._name = name
    this._code = code
    this._live = live
    this.initializeSorting()
  }

  get live () {
    return this._live
  }

  get name () {
    return this._name
  }

  get code () {
    return this._code
  }

  get sortKey () {
    return this._sortKey
  }

  get sortCategory () {
    return this._sortCategory
  }

  initializeSorting () {
    let name = this._name.toLowerCase()
    const prefix = find(IGNORED_PREFIXES, (pre) => name.startsWith(pre + ' '))

    if (prefix) {
      name = name.substring(prefix.length + 1).trim()
    }

    this._sortKey = name
    this._sortCategory = isEmpty(name) ? '?' : name[0].toUpperCase()
  }
}
