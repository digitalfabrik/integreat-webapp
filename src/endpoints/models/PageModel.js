class PageModel {
  constructor ({id, numericId, title = '', parentNumericId = null, content = '', thumbnail = null, order = 0, availableLanguages = {}, url = null}) {
    this._id = id
    this._numericId = numericId
    this._title = title
    this._content = content
    this._parentNumericId = parentNumericId
    this._thumbnail = thumbnail
    this._order = order
    this._availableLanguages = availableLanguages
    this._children = []
    this._parent = null
    this._url = url
  }

  get thumbnail () {
    return this._thumbnail
  }

  addChild (page) {
    this._children.push(page)
    this._children.sort((page1, page2) => page1.order - page2.order)
  }

  setParent (parent) {
    this._parent = parent
  }

  get parent () {
    return this._parent
  }

  get id () {
    return this._id
  }

  get numericId () {
    return this._numericId
  }

  get title () {
    return this._title
  }

  get content () {
    return this._content
  }

  get parentNumericId () {
    return this._parentNumericId
  }

  get children () {
    return this._children
  }

  get order () {
    return this._order
  }

  get availableLanguages () {
    return this._availableLanguages
  }

  get url () {
    if (this._url) {
      return this._url
    }
    return `${this.parent.url}/${this.id}`
  }
}

export const EMPTY_PAGE = new PageModel({})
export default PageModel
