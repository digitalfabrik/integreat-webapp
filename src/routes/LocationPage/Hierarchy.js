import { EMPTY_PAGE } from 'endpoints/models/PageModel'

import normalizeUrl from 'normalize-url'

import { last } from 'lodash/array'
import { isEmpty } from 'lodash/lang'
import { forEach } from 'lodash/collection'

class Hierarchy {
  constructor (path = '') {
    this._code = path ? path.split('/').filter((path) => path !== '') : []
    this._pages = []
  }

  /**
   * Finds the current page which should be rendered based on {@link this.props.path}
   * The page can be accessed by calling {@link Hierarchy.top()}
   * @return {*} The model to renders
   */
  build (rootPage) {
    let error = null

    if (!rootPage || rootPage === EMPTY_PAGE) {
      return error
    }

    let currentPage = rootPage

    let pages = [currentPage]

    forEach(this._code, (id) => {
      currentPage = currentPage.children.find((page) => page.id === id)

      if (!currentPage) {
        error = 'errors:page.notFound'
        return false
      }

      pages.push(currentPage)
    })

    this._pages = pages

    return error
  }

  /**
   * Maps the pages which represent this hierarchy
   *
   * @param fn The mapper function which takes the page and the path
   */
  map (fn) {
    let path = ''
    return this.pages.map((page) => {
      if (page.numericId !== 0) {
        path += '/' + page.id
      }
      return fn(page, path)
    })
  }

  path () {
    return normalizeUrl('/' + this._code.join('/'), {removeTrailingSlash: true})
  }

  get pages () {
    return this._pages
  }

  /**
   * If the path is "/1/5/6", then the page with the id 6 is returned
   * @returns {*} The top element in this hierarchy
   */
  top () {
    return last(this._pages)
  }

  /**
   * @returns {*} true if this hierarchy is empty
   */
  root () {
    return isEmpty(this._code)
  }
}

export default Hierarchy
