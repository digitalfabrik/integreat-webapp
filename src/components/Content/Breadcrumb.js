import React from 'react'
import PropTypes from 'prop-types'

import style from './Breadcrumb.css'
import Hierarchy from 'routes/LocationPage/Hierarchy'
import { Link } from 'redux-little-router'
import LocationModel from '../../endpoints/models/LocationModel'
import { LocationFetcher } from '../../endpoints/index'

class BreadcrumbAdapter extends React.Component {
  static propTypes = {
    location: PropTypes.string.isRequired,
    language: PropTypes.string.isRequired,
    hierarchy: PropTypes.instanceOf(Hierarchy).isRequired,
    locations: PropTypes.arrayOf(PropTypes.instanceOf(LocationModel))
  }

  getFormattedTitle (page) {
    const location = this.props.locations.find((location) => location.code === page.title)

    if (location) {
      return location.name
    }

    return null
  }

  render () {
    const hierarchy = this.props.hierarchy
    return (
      <div className={style.breadcrumbs}>{hierarchy.map((page, path) => {
        return (
          <Link key={page.id} className={style.breadcrumb}
                href={`/${this.props.location}/${this.props.language}${path}`}>
            <span className={style.separator}/>
            <span className={style.level}>{this.getFormattedTitle(page) || page.title}</span>
          </Link>
        )
      })}
      </div>
    )
  }
}

class Breadcrumb extends React.Component {
  static propTypes = {
    hierarchy: PropTypes.instanceOf(Hierarchy).isRequired,
    location: PropTypes.string.isRequired,
    language: PropTypes.string.isRequired
  }

  render () {
    return <LocationFetcher hideSpinner={true}>
      <BreadcrumbAdapter hierarchy={this.props.hierarchy} language={this.props.language}
                         location={this.props.location}/>
    </LocationFetcher>
  }
}

export default Breadcrumb
