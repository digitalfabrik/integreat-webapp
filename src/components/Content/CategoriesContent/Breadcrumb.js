import React from 'react'
import PropTypes from 'prop-types'

import { Link } from 'redux-little-router'

import Hierarchy from 'routes/LocationPage/Hierarchy'
import LOCATIONS_ENDPOINT from 'endpoints/location'
import withFetcher from 'endpoints/withFetcher'

import style from './Breadcrumb.css'

class Breadcrumb extends React.Component {
  static propTypes = {
    hierarchy: PropTypes.instanceOf(Hierarchy).isRequired,
    location: PropTypes.string.isRequired,
    language: PropTypes.string.isRequired
  }

  getFormattedTitle (page) {
    const location = this.props.locations.find((location) => location.code === page.title)

    if (location) {
      return location.name
    }

    return null
  }

  render () {
    return <div className={style.breadcrumbs}>{this.props.hierarchy.map((page, path) => {
      if (page !== this.props.hierarchy.top()) {
        return (
          <Link key={page.id} className={style.breadcrumb}
                href={`/${this.props.location}/${this.props.language}${path}`}>
            <span className={style.separator}/>
            <span className={style.level}>{this.getFormattedTitle(page) || page.title}</span>
          </Link>
        )
      }
    })}
    </div>
  }
}

export default withFetcher(LOCATIONS_ENDPOINT, true, true)(Breadcrumb)
