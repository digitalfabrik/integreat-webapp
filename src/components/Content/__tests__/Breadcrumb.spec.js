import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { Provider } from 'react-redux'

import { Breadcrumb } from '../Breadcrumb'
import PageModel from '../../../endpoints/models/PageModel'
import Hierarchy from 'routes/LocationPage/Hierarchy'
import LocationModel from '../../../endpoints/models/LocationModel'
import store from '../../../store'

describe('Breadcrumb', () => {
  const page4 = new PageModel({
    id: 'page4',
    numericId: 4,
    title: 'Page4',
    parent: 2
  })

  const page2 = new PageModel({
    id: 'page2',
    numericId: 2,
    title: 'Page2',
    parent: 0,
    children: [page4]
  })

  const page3 = new PageModel({
    id: 'page3',
    numericId: 3,
    title: 'Page3',
    parent: 0
  })

  const rootPage = new PageModel({
    id: 'rootId',
    numericId: 0,
    title: 'augsburg',
    children: [page2, page3]
  })

  const locations = [
    new LocationModel({
      name: 'Augsburg',
      code: 'augsburg'
    }),
    new LocationModel({
      name: 'Other city',
      code: 'otherCity'
    })]

  let hierarchy = new Hierarchy('augsburg/page2/page4')
  hierarchy.build(rootPage)

  test('should render', () => {
    const wrapper = renderer.create(
      <Provider store={store}>
        <Breadcrumb
          hierarchy={hierarchy}
          location={'augsburg'}
          language={'de'}
          locations={locations}/>
      </Provider>)

    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  test('getFormattedTitle', () => {
    const component = shallow(
      <Breadcrumb
        hierarchy={hierarchy}
        location={'augsburg'}
        language={'de'}
        locations={locations}/>)
      .instance()

    expect(component.getFormattedTitle(rootPage)).toBe('Augsburg')
    expect(component.getFormattedTitle(page2)).toBe(null)
  })
})
