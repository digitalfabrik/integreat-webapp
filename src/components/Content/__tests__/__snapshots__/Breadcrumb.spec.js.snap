// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb should render 1`] = `
<div
  className="breadcrumbs"
/>
`;
